staging/src/k8s.io/cloud-provider/app/webhooks.go

```diff
@@ -0,0 +1,184 @@
+	"k8s.io/apimachinery/pkg/runtime"
+	"k8s.io/apimachinery/pkg/runtime/serializer"
+	"k8s.io/apimachinery/pkg/util/sets"
+	"k8s.io/apiserver/pkg/server/mux"
+	cloudprovider "k8s.io/cloud-provider"
+	"k8s.io/cloud-provider/app/config"
+	genericcontrollermanager "k8s.io/controller-manager/app"
+	"k8s.io/klog/v2"
+)
+
+var (
+	runtimeScheme = runtime.NewScheme()
+	codecs        = serializer.NewCodecFactory(runtimeScheme)
+	deserializer  = codecs.UniversalDeserializer()
+	encoder       runtime.Encoder
+)
+
+func init() {
+	_ = corev1.AddToScheme(runtimeScheme)
+	_ = admissionregistrationv1.AddToScheme(runtimeScheme)
+	_ = admissionregistrationv1beta1.AddToScheme(runtimeScheme)
+	_ = admissionv1.AddToScheme(runtimeScheme)
+	_ = admissionv1beta1.AddToScheme(runtimeScheme)
+	serializerInfo, _ := runtime.SerializerInfoForMediaType(codecs.SupportedMediaTypes(), runtime.ContentTypeJSON)
+	encoder = serializerInfo.Serializer
+}
+
+// WebhooksDisabledByDefault is the webhooks disabled default when starting cloud-controller managers.
+var WebhooksDisabledByDefault = sets.NewString()
+
+type WebhookConfig struct {
+	Path             string
+	AdmissionHandler func(*admissionv1.AdmissionRequest) (*admissionv1.AdmissionResponse, error)
+}
+
+type webhookHandler struct {
+	path string
+	http.Handler
+	admissionHandler func(*admissionv1.AdmissionRequest) (*admissionv1.AdmissionResponse, error)
+	completedConfig  *config.CompletedConfig
+	cloud            cloudprovider.Interface
+}
+
+func newWebhookHandlers(webhookConfigs map[string]WebhookConfig, completedConfig *config.CompletedConfig, cloud cloudprovider.Interface) map[string]webhookHandler {
+	webhookHandlers := make(map[string]webhookHandler)
+	webhooksEnabled := ""
+	for name, config := range webhookConfigs {
+		if !genericcontrollermanager.IsControllerEnabled(name, WebhooksDisabledByDefault, completedConfig.ComponentConfig.Webhook.Webhooks) {
+			klog.Warningf("Webhook %q is disabled", name)
+			continue
+		}
+		webhooksEnabled += name
+		webhookHandlers[name] = webhookHandler{
+			path:             config.Path,
+			admissionHandler: config.AdmissionHandler,
+			completedConfig:  completedConfig,
+			cloud:            cloud,
+		}
+	}
+	klog.Infof("Webhooks enabled: %q", webhooksEnabled)
+	return webhookHandlers
+}
+
+func WebhookNames(webhooks map[string]WebhookConfig) []string {
+	ret := sets.StringKeySet(webhooks)
+	return ret.List()
+}
+
+func newHandler(webhooks map[string]webhookHandler) *mux.PathRecorderMux {
+	mux := mux.NewPathRecorderMux("controller-manager-webhook")
+
+	for _, handler := range webhooks {
+		mux.Handle(handler.path, handler)
+	}
+
+	return mux
+}
+
+func (h webhookHandler) ServeHTTP(w http.ResponseWriter, r *http.Request) {
+	klog.Infof("Received validation request: %q", r.RequestURI)
+
+	in, err := parseRequest(r)
+	if err != nil {
+		klog.Error(err)
+		http.Error(w, err.Error(), http.StatusBadRequest)
+		return
+	}
+
+	admissionResponse, err := h.admissionHandler(in.Request)
```

Comment:
We should probably wrap these requests in metrics. I would think we would want latency and response code metrics by canoncial webhook name.